components/schema-migrator/migrations/director/20211122114401_storage-redesign.up.sql

```diff
@@ -0,0 +1,714 @@
+BEGIN;
+
+CREATE TABLE tenant_applications
+(
+    tenant_id   uuid NOT NULL,
+    id uuid NOT NULL,
+    owner       bool,
+
+    FOREIGN KEY (id) REFERENCES applications (id) ON DELETE CASCADE,
+    FOREIGN KEY (tenant_id) REFERENCES business_tenant_mappings (id),
+    PRIMARY KEY (tenant_id, id)
+);
+
+CREATE TABLE tenant_runtimes
+(
+    tenant_id   uuid NOT NULL,
+    id uuid NOT NULL,
+    owner       bool,
+
+    FOREIGN KEY (id) REFERENCES runtimes (id) ON DELETE CASCADE,
+    FOREIGN KEY (tenant_id) REFERENCES business_tenant_mappings (id),
+    PRIMARY KEY (tenant_id, id)
+);
+
+CREATE OR REPLACE FUNCTION on_insert_new_parent() RETURNS TRIGGER AS
+$$
+DECLARE
+    tenant_table TEXT;
+BEGIN
+    IF NEW.parent IS NOT NULL
+    THEN
+        FOREACH tenant_table IN ARRAY TG_ARGV -- for each top level resource
+            LOOP
+                -- mark him as owner to everything that is owned by its child
+                EXECUTE format('INSERT INTO %I (SELECT %L, id, owner FROM %I WHERE owner = true AND tenant_id = %L) ON CONFLICT DO NOTHING;', tenant_table, NEW.parent, tenant_table, NEW.id);
+            END LOOP;
+    END IF;
+    IF OLD.parent IS NOT NULL AND NEW.parent <> OLD.parent
+    THEN
+        FOREACH tenant_table IN ARRAY TG_ARGV -- for each top level resource
+            LOOP
+                -- delete all accesses from the old parent that are for resources of the moved tenant
+                EXECUTE format('DELETE FROM %I WHERE id IN (SELECT id FROM %I WHERE tenant_id = %L AND owner = true) AND tenant_id = %L AND owner = true', tenant_table, tenant_table, NEW.id, OLD.parent);
+            END LOOP;
+    END IF;
+    RETURN NULL;
+END
+$$ LANGUAGE plpgsql;
+
+CREATE TRIGGER insert_parent AFTER INSERT OR UPDATE ON business_tenant_mappings FOR EACH ROW EXECUTE PROCEDURE on_insert_new_parent('tenant_applications', 'tenant_runtimes');
+
+-- this trigger replaces the cascade on tenant_id foreign key in order to execute it BEFORE the tenant is deleted.
+-- that is needed because in order for delete_parent_chain select parent to work (before the tenant is actually deleted)
+CREATE OR REPLACE FUNCTION on_delete_tenant() RETURNS TRIGGER AS
```

Comment:
If we remove the delete_parent_chain trigger, yes. If we leave it, no because the cascade is triggered AFTER delete and the select from business_tenant_mappings in delete_parent_chain fail. Therefore we need the cascade to happen BEFORE the actual delete -> this trigger.

Reply:
ok so maybe it makes sense to remove the delete trigger i think, together with it this one goes away as well as the delete_resource one, right ?